src/bsoncxx/test/optional.test.cpp

```diff
@@ -0,0 +1,238 @@
+                  "fail");
+    static_assert(std::is_constructible<To1, From1 const&&>::value  //
+                      == std::is_constructible<To2, From2 const&&>::value,
+                  "fail");
+    return true;
+}
+
+template <typename From, typename To>
+bool check_conversions() {
+    return check_convert_alike<From, To, optional<From>, optional<To>>() &&
+           check_convert_alike<From, To, From, optional<To>>() &&
+           check_construct_alike<From, To, optional<From>, optional<To>>() &&
+           check_construct_alike<From, To, From, optional<To>>();
+}
+
+template <typename T>
+bool static_checks() {
+    assert_sameness<std::is_copy_constructible, T>();
+    assert_sameness<std::is_copy_assignable, T>();
+    assert_sameness<std::is_move_constructible, T>();
+    assert_sameness<std::is_move_assignable, T>();
+    assert_sameness<bsoncxx::detail::is_equality_comparable, T>();
+    assert_sameness<bsoncxx::detail::is_totally_ordered, T>();
+    static_assert(bsoncxx::detail::is_equality_comparable<T, optional<T>>::value ==
+                      bsoncxx::detail::is_equality_comparable<T>::value,
+                  "fail");
+    static_assert(bsoncxx::detail::is_totally_ordered_with<T, optional<T>>::value ==
+                      bsoncxx::detail::is_totally_ordered<optional<T>>::value,
+                  "fail");
+    static_assert(
+        std::is_constructible<optional<T>, T>::value == std::is_constructible<T, T>::value, "fail");
+    static_assert(std::is_constructible<optional<T>, bsoncxx::stdx::nullopt_t>{}, "fail");
+    // Assert we return proper reference types
+    static_assert(std::is_same<deref_t<optional<T>>, T&&>{}, "fail");
+    static_assert(std::is_same<deref_t<optional<T> const>, const T&&>{}, "fail");
+    static_assert(std::is_same<deref_t<optional<T> const&>, const T&>{}, "fail");
+    static_assert(std::is_same<deref_t<optional<T>&>, T&>{}, "fail");
+    // .value()
+    static_assert(std::is_same<value_t<optional<T>>, T&&>{}, "fail");
+    static_assert(std::is_same<value_t<optional<T> const>, const T&&>{}, "fail");
+    static_assert(std::is_same<value_t<optional<T> const&>, const T&>{}, "fail");
+    static_assert(std::is_same<value_t<optional<T>&>, T&>{}, "fail");
+    // operator->
+    static_assert(std::is_same<arrow_t<optional<T>>, T*>{}, "fail");
+    static_assert(std::is_same<arrow_t<optional<T> const>, const T*>{}, "fail");
+    static_assert(std::is_same<arrow_t<optional<T> const&>, const T*>{}, "fail");
+    static_assert(std::is_same<arrow_t<optional<T>&>, T*>{}, "fail");
+    return check_conversions<T, T>();
+}
+
+}  // namespace
+
+static_assert(bsoncxx::detail::is_totally_ordered<std::string>{}, "fail");
+static_assert(bsoncxx::detail::is_totally_ordered<int>{}, "fail");
+static_assert(!bsoncxx::detail::is_totally_ordered<not_ordered>{}, "fail");
+
+TEST_CASE("Trait checks") {
+    CHECK(static_checks<int>());
```

Comment:
Check fails on the recently added VS 2019 tasks. [Example](https://spruce.mongodb.com/task/cxx_driver_integration_vs2019_4.0_single_compile_and_test_with_shared_libs_extra_alignment_patch_5b743b105e13c4437b18d8245db8e3d83c351c88_65a77501a4cf47738e03849b_24_01_17_06_34_42/logs?execution=0):

```
[2024/01/17 01:41:48.803] C:\data\mci\365e5f1b9d6299dcc3ea13c579d03f8d\mongo-cxx-driver\src\bsoncxx\test\optional.test.cpp(53,35): error C2338: Fail [C:\data\mci\365e5f1b9d6299dcc3ea13c579d03f8d\mongo-cxx-driver\build\src\bsoncxx\test\test_bson.vcxproj]
[2024/01/17 01:41:48.803] C:\data\mci\365e5f1b9d6299dcc3ea13c579d03f8d\mongo-cxx-driver\src\bsoncxx\test\optional.test.cpp(112): message : see reference to function template instantiation 'bool `anonymous-namespace'::assert_sameness<bsoncxx::detail::is_totally_ordered,T>(void)' being compiled [C:\data\mci\365e5f1b9d6299dcc3ea13c579d03f8d\mongo-cxx-driver\build\src\bsoncxx\test\test_bson.vcxproj]
[2024/01/17 01:41:48.803]           with
[2024/01/17 01:41:48.803]           [
[2024/01/17 01:41:48.803]               T=int
[2024/01/17 01:41:48.803]           ]
[2024/01/17 01:41:48.803] C:\data\mci\365e5f1b9d6299dcc3ea13c579d03f8d\mongo-cxx-driver\src\bsoncxx\test\optional.test.cpp(147): message : see reference to function template instantiation 'bool `anonymous-namespace'::static_checks<int>(void)' being compiled [C:\data\mci\365e5f1b9d6299dcc3ea13c579d03f8d\mongo-cxx-driver\build\src\bsoncxx\test\test_bson.vcxproj]
```

Reply:
Appears resolved in bf1a6b024416a12c76fa41291ad0e81f17ba9502